### PR TITLE
Export: solved my_clippings OOM

### PR DESCRIPTION
**Pull request by Mochitto**
This should take care of #10745 regarding my_clippings; I'm not sure if it's the last one remaining and the issue can be closed.
The result is the same and I've followed the instructions given by @pazos; hopefully the OOM risk is no more.

Thanks for tagging me in #12106 (probably needs to solve conflicts with this fix)

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/12107)

### CODE REVIEW

**Comment by hius07** [comment]
I'd propose building each line and then `table.concat(tbl, "\n")` to avoid extra concatenation.

File: plugins/exporter.koplugin/target/my_clippings.lua [hunk 1]
```diff
@@ -11,24 +11,35 @@ local ClippingsExporter = require("base"):new {
...
+
+                    table.insert(tbl, clipping.note)
+                    table.insert(tbl, "\n==========\n")
                 end
             end
         end
     end
+
+    local content = table.concat(tbl)
     return content
```
Comment (Frenzie) [inline plugins/exporter.koplugin/target/my_clippings.lua:43]: Reads a bit funny. ;-)

Comment (Mochitto) [reply plugins/exporter.koplugin/target/my_clippings.lua:43]: Oh gosh ahah thanks for pointing it out!

**Comment by Mochitto** [comment]
@hius07 

I've tried using the concat "\n" but the result is inconsistent..?
I'm not sure why that would be so, since I'm matching the same structure as the one with the concat "\n"

Can you help me out? I'm not very familiar with lua
```
local function format(booknotes)
    local tbl = {}

    for ___, entry in ipairs(booknotes) do
        for ____, clipping in ipairs(entry) do
            if booknotes.title and clipping.text then
                table.insert(tbl, booknotes.title)

                local header = T(_("- Your highlight on page %1 | Added on %2"), clipping.page,
                    os.date("%A, %B %d, %Y %I:%M:%S %p", clipping.time))
                table.insert(tbl, header)

                table.insert(tbl, "")
                table.insert(tbl, clipping.text)
                table.insert(tbl, "==========")

                if clipping.note then
                    table.insert(tbl, booknotes.title)

                    header = T(_("- Your note on page %1 | Added on %2"), clipping.page,
                        os.date("%A, %B %d, %Y %I:%M:%S %p", clipping.time))
                    table.insert(tbl, header)

                    table.insert(tbl, "")
                    table.insert(tbl, clipping.note)
                    table.insert(tbl, "==========")
                end
            end
        end
    end

    local content = table.concat(tbl, "\n")
    return content
end
```

The outputs; notice how some titles are in the same line as the divider `==========`
```
My experience with Anki
- Your highlight on page 1 | Added on Thursday, June 27, 2024 08:38:52 PM

I prepared some questions that are primarily for reflection, feel free
==========
My experience with Anki
- Your highlight on page 1 | Added on Thursday, June 27, 2024 08:38:54 PM

Do you have a special philosophy/methodology behind
==========
My experience with Anki
- Your highlight on page 6 | Added on Thursday, June 27, 2024 08:38:58 PM

ease hell](https://readbroca.com/anki/ease-hell/)
- Progress
==========KOReader Quickstart Guide
- Your highlight on page 1 | Added on Thursday, June 27, 2024 08:35:36 PM

You can access the complete user manual from
==========LinuxCommand.Find and replace
- Your highlight on page 1 | Added on Thursday, June 27, 2024 08:39:19 PM

-rli "1.0.5" . | xargs -i@ sed -i "s/1.0
==========
LinuxCommand.Find and replace
- Your highlight on page 1 | Added on Thursday, June 27, 2024 08:39:57 PM

the matching string with [[LinuxTool.grep]].
`r` recursively, `l` print name of files with matching strings, `i`
==========
LinuxCommand.Find and replace
- Your note on page 1 | Added on Thursday, June 27, 2024 08:39:57 PM

This is my note
==========

```

**Comment by Frenzie** [comment]
At the beginning insert an extra `"\n"` (so it'll turn into two).

**Comment by Frenzie** [comment]
That is, if you do concat with \n you get `line1\nline2\nline3`. Attach that to another set and you're missing a newline.

**Comment by Mochitto** [comment]
By adding a new `table.insert(tbl, "")` in the beginning, there's extra newlines after the divider.

I don't really understand why only *some* titles have the extra newline and some do not. It doesn't even have to do with the difference between notes and highlights.
This doesn't happen with the newlines and without the concat.

Output with the newline/insert in the beginning
```
LinuxCommand.Find and replace
- Your highlight on page 1 | Added on Thursday, June 27, 2024 08:39:19 PM

-rli "1.0.5" . | xargs -i@ sed -i "s/1.0
==========

LinuxCommand.Find and replace
- Your highlight on page 1 | Added on Thursday, June 27, 2024 08:39:57 PM

the matching string with [[LinuxTool.grep]].
`r` recursively, `l` print name of files with matching strings, `i`
==========
LinuxCommand.Find and replace
- Your note on page 1 | Added on Thursday, June 27, 2024 08:39:57 PM

This is my note
==========
KOReader Quickstart Guide
- Your highlight on page 1 | Added on Thursday, June 27, 2024 08:35:36 PM

You can access the complete user manual from
==========
My experience with Anki
- Your highlight on page 1 | Added on Thursday, June 27, 2024 08:38:52 PM

I prepared some questions that are primarily for reflection, feel free
==========

My experience with Anki
- Your highlight on page 1 | Added on Thursday, June 27, 2024 08:38:54 PM

Do you have a special philosophy/methodology behind
==========

My experience with Anki
- Your highlight on page 6 | Added on Thursday, June 27, 2024 08:38:58 PM

ease hell](https://readbroca.com/anki/ease-hell/)
- Progress
==========
```

Tysm for the help

**Comment by Mochitto** [comment]
Oh, I've got it now! It actually needed to be added to the end part :)
I lost focus of the fact that the format function is called separately for each book, instead of the entry i'm iterating through being a continuous stream.

tysm, fixed ! :)

File: plugins/exporter.koplugin/target/my_clippings.lua [hunk 2]
```diff
@@ -11,25 +11,41 @@ local ClippingsExporter = require("base"):new {
...
+                    table.insert(tbl, "==========")
                 end
             end
         end
     end
-    return content
+
+    -- Add an extra newline to keep a consistent format when exporting
+    -- multiple books' data; if this is missing, the title is written
+    -- after the previous book's last "=========" as "==========New book title"
```
Comment (Frenzie) [inline plugins/exporter.koplugin/target/my_clippings.lua:46]: I think it's fairly self-explanatory, but if we feel a comment is in order I'd just keep it brief. Something like:

```suggestion
    -- Ensure a newline after the previous book's last "=========".
```

**Review by pazos — APPROVED** [review]
LGTM!, thank you very much :)

**Comment by arpanghosh8453** [comment]
@Mochitto and @pazos In order to include the features in #12106 (now with conflicting changes) you need to change the following in line `19` (OOM fix version). I have added the author as "Unknown" if the author is not presenting preserving Kindle syntax as mentioned in @hius07 in #12106 

```diff
- table.insert(tbl, booknotes.title)

+ if booknotes.author then
+     table.insert(tbl, booknotes.title .. " (" .. booknotes.author .. ")" )
+ else
+     table.insert(tbl, booknotes.title .. " (Unknown)" )
+ end
```

And the following in line `30` (OOM fix version)

```diff
- table.insert(tbl, booknotes.title)

+ if booknotes.author then
+     table.insert(tbl, booknotes.title .. " (" .. booknotes.author .. ")" )
+ else
+     table.insert(tbl, booknotes.title .. " (Unknown)" )
+ end
```

**Comment by arpanghosh8453** [comment]
Also @Mochitto and @pazos, I have noticed something, that is worth fixing. The time in the added export are in local time. But I assume Kindle uses GMT/UTC timestamps so when importing to myclippings, it auto-converts them to local time based on browser timezone/system time. I have illustrated an example here - 

image

image

As you can see, 05:30 was added to the timestamp given in the file ( I am from GMT + 5:30 time zone )

I am not sure what is the format of the time in `clipping.time` , but it's best if you can somehow convert that to the UTC time zone matching Kindle syntax for seamless importing by clippings.io

**Comment by hius07** [comment]
> I am not sure what is the format of the time in `clipping.time` , but it's best if you can somehow convert that to the UTC time zone matching Kindle syntax for seamless importing by clippings.io.

It is converted exactly to the Kindle format:
https://github.com/koreader/koreader/blob/6b81b5190fb1d7d63b8f2f4b89e6bf4cb08271b0/plugins/exporter.koplugin/target/my_clippings.lua#L21-L22

**Comment by arpanghosh8453** [comment]
> > I am not sure what is the format of the time in `clipping.time` , but it's best if you can somehow convert that to the UTC time zone matching Kindle syntax for seamless importing by clippings.io.
> 
> It is converted exactly to the Kindle format:
> 
> https://github.com/koreader/koreader/blob/6b81b5190fb1d7d63b8f2f4b89e6bf4cb08271b0/plugins/exporter.koplugin/target/my_clippings.lua#L21-L22

The problem is outlined above. What I meant is I do not know what object `clipping.time` variable is ( is it a formatted string, datetime object, Epoch time integer ... ) . The format there is fine, but the time is in local time zone. Which is not how Kindle exports it. So when uploaded to myclippings.io (as intended for this plugin by @Mochitto), the timestamp gets modified to wrong value ( It always assumes the given time is in UTC and stores accordingly, shows the user in local time by calculating from browser time ). See the images I attached in my comment. That might help you understand the issue.

**Comment by hius07** [comment]
What does myclippings.io import from the Kindle?
My Clippings.txt or something else?

**Comment by arpanghosh8453** [comment]
> What does myclippings.io import from the Kindle? My Clippings.txt or something else?

Yes, It imports the My Clippings.txt file (The exact format we are trying to replicate here)

**Comment by hius07** [comment]
And our Export plugin provides the same timestamp as the Kindle does.
Maybe myclippings.io wants to connect to the device to get its timezone and to recalculate the time?

**Comment by hius07** [comment]
Okay, I got your point.
I am looking at the My Clippings.txt made just now.
It contains the timestamp exactly with the current time of my Kindle. It is UTC+2.

**Comment by arpanghosh8453** [comment]
> Okay, I got your point. I am looking at the My Clippings.txt made just now. It contains the timestamp exactly with the current time of my Kindle. It is UTC+2.

Ah, Thank you. So you mean it's the local time and not UTC. Do you mind uploading that to the service ( you just need a simple https://clippings.io account ) to see if the time appears right?

**Comment by Mochitto** [comment]
I haven't used apps that integrate with my clippings, only obsidian plugins which worked fine; I just made the plugin since kindle has a huge market-share and thought it would make for a very compatibility with other projects.

I'll wait for the updates but my 2 cents are: I don't think kindle end-users would know what timestamps other than local timestamps would be.
I'm rather perplexed about the way this service handles time, as we are just exploring a timezone-less datetime string...
So any parsing should be pretty forward and not apply mutations; if kindle devices export in local timezones (which I'd expect, else I could imagine all the calls to Amazon's support services), could it be that the app applies the timezones erroneously?

Maybe you can shoot an email to the app's support services and ask about it :)

**Comment by Mochitto** [comment]
Maybe a new issue can be opened about this as it's out of scope with the out of memory fix and the authors fix.

**Comment by arpanghosh8453** [comment]
> I haven't used apps that integrate with my clippings, only obsidian plugins which worked fine; I just made the plugin since kindle has a huge market-share and thought it would make for a very compatibility with other projects.
> 
> I'll wait for the updates but my 2 cents are: I don't think kindle end-users would know what timestamps other than local timestamps would be. I'm rather perplexed about the way this service handles time, as we are just exploring a timezone-less datetime string... So any parsing should be pretty forward and not apply mutations; if kindle devices export in local timezones (which I'd expect, else I could imagine all the calls to Amazon's support services), could it be that the app applies the timezones erroneously?
> 
> Maybe you can shoot an email to the app's support services and ask about it :)

Then I assume that is the issue with the service itself. Don't bother about it. For now, I will be happy with merging #12106 and #12107 (both looks approved, but have conflicting changes). I am not sure how @pazos or @Frenzie will resolve the conflict, but in case of any issue, I am down to help.

**Pull request merged**